syntax = "proto3";

package chat;

service ChatService {
    rpc Login(User) returns (LoginResponse) {}
    rpc Logout(User) returns (LoginResponse) {}
    rpc CreateAccount(User) returns (CreateAccountResponse) {}
    rpc DeleteAccount(User) returns (DeleteAccountResponse) {}
    rpc ListAccounts(ListAccountsRequest) returns (ListAccountsResponse) {}
    rpc SendMessage(Message) returns (stream SendMessageResponse) {}
}

message User {
    string username = 1;
}

message UserList {
    repeated User users = 1;
}

message LoginResponse {
    bool success = 1;
    string message = 2;
}


message CreateAccountResponse {
    bool success = 1;
    string message = 2;
}

message DeleteAccountResponse {
    bool success = 1;
    string message = 2;
}

message ListAccountsRequest {
    string pattern = 1;
}

message ListAccountsResponse {
    repeated string usernames = 1;
}

message Message {
    string sender = 1;
    string receiver = 2;
    string text = 3;
}

message SendMessageResponse {
    bool success = 1;
    string message = 2;
}
